src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs

```diff
@@ -580,5 +600,86 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+        /// </summary>
+        private class PossiblyImplicitlyTypedDeconstructionLocalSymbol : SourceLocalSymbol
+        {
+            public PossiblyImplicitlyTypedDeconstructionLocalSymbol(
+                Symbol containingSymbol,
+                Binder binder,
+                TypeSyntax typeSyntax,
+                SyntaxToken identifierToken,
+                LocalDeclarationKind declarationKind)
+            : base(containingSymbol, binder, RefKind.None, typeSyntax, identifierToken, declarationKind)
+            {
+#if DEBUG
+                SyntaxNode parent = IdentifierToken.Parent;
+                Debug.Assert(parent != null);
+                Debug.Assert(parent.Kind() == SyntaxKind.VariableDeclarator);
+                parent = parent.Parent;
+                Debug.Assert(parent != null);
+                Debug.Assert(parent.Kind() == SyntaxKind.VariableDeclaration);
+
+                // The parents should be alternating VariableDeconstructionDeclarator and VariableDeclaration, until a different kind is found
+                while (true)
+                {
+                    parent = parent.Parent;
+                    Debug.Assert(parent != null);
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator)
+                    {
+                        break;
+                    }
+
+                    parent = parent.Parent;
+                    Debug.Assert(parent != null);
+                    Debug.Assert(parent.Kind() == SyntaxKind.VariableDeclaration);
+                }
+
+                Debug.Assert(parent.Kind() == SyntaxKind.LocalDeclarationStatement);
+#endif
+            }
+
+            protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+            {
+                // Try binding immediately enclosing deconstruction-declaration (the top-level VariableDeclaration), this should force the inference.
+
+                var topLevelVariableDeclaration = IdentifierToken.
+                                                    Parent. // VariableDeclaratorSyntax
+                                                    Parent; // VariableDeclarationSyntax
+                while (true)
+                {
+                    var parent = topLevelVariableDeclaration.Parent;
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator)
+                    {
+                        break;
+                    }
+
+                    var grandParent = parent.Parent;
+                    Debug.Assert(grandParent.Kind() == SyntaxKind.VariableDeclaration);
+
+                    topLevelVariableDeclaration = grandParent;
+                }
+
+                Debug.Assert(((VariableDeclarationSyntax)topLevelVariableDeclaration).Deconstruction != null);
+                Debug.Assert(((VariableDeclarationSyntax)topLevelVariableDeclaration).Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
```

Comment:
Perhaps extract a `GetLocalDeclarationStatement(IdentifierToken)` helper method that is used here and in the `.ctor` above and have that helper method assert there at least one pair of { `VariableDeconstructionDeclarator`, `VariableDeclaration` }.